marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/InputEndpointImpl.java

```diff
@@ -208,24 +214,22 @@ private void processInput(CallContext callContext) {
 			public Boolean call() {
 				try {
 					CallContext callContext = bulkInputCallerImpl.getCallContexts().poll();
+					if(callContext != null)
+						bulkInputCallerImpl.processInput(callContext);
 
-					bulkInputCallerImpl.processInput(callContext);
 					if(getPhase() == WorkPhase.COMPLETED && bulkInputCallerImpl.getCallContexts().isEmpty()) {
 						getCallerThreadPoolExecutor().shutdown();
-					} else {
+					}
+					else if(callContext != null) {
 						bulkInputCallerImpl.getCallContexts().put(callContext);
+						submitTask(this);
```

Comment:
By contrast with the output case, the input case is driven by the application adding input.

So, the callable shouldn't resubmit itself but, instead, should shutdown the thread pool if the thread pool is awaiting termination and the task queue is empty.